### ISSUE TITLE
Custom layer input tensor tracking support in profiler

### ISSUE THREAD

**Issue by alexmsettle** [pytorch/pytorch#82344]
### 🚀 The feature, motivation and pitch

This is related to a previous feature request #66105.  We updated the profiler to include logic that tracks the ops that produce the input tensors to downstream ops.  This information can be used to construct the network graph.  Recently I noticed that the input tensor tracking doesn't work for the case where there is a custom pytorch layer (module) which is implemented with custom kernels.  If the custom kernels do not use aten:: ops then the existing tensor tracking software does not get activated.  This use case is common for engineers working on MLPerf networks.  Typically an engineer will write a cuda kernel, then use the custom layer recipe to register it as a PyTorch module.  We would like to be able to construct the network graph for these cases.

The code that would need to be modified is 
  THPFunction_apply()   in python_fuction.cpp

https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_function.cpp#L753

I implemented this feature in my copy of pytorch but I have not done much testing on it yet.

### Alternatives

I don't think there is a good workaround if you want to generate a network graph that includes the custom layer with custom kernel.  I think you would have to register the kernel with the aten lib which seems complicated.

### Additional context

![image](https://user-images.githubusercontent.com/37422826/181349234-927e1779-4941-4a12-9c29-f4f0e6d7ac81.png)

This trace is an example of what it looks like when the feature works.  The CustomLayer has op_id = 25.  This tensor produced by this layer feeds the softmax layer.  You can see this from input_op_ids = [(25,0)].  This means the input tensor was produced by op_id 25.  Without the fix, the input_op_ids to the softmax layer are unknown.

cc @ilia-cher @robieta @chaekit @gdankel @bitfort @ngimel @nbcsm @guotuofeng @guyang3532 @gaoteng-git @VitalyFedyunin

**Comment by ngimel** [comment]
You need to register your custom operators with TORCH_LIBRARY so that they participate in all dispatcher-based activities https://pytorch.org/cppdocs/library.html

**Comment by alexmsettle** [comment]
Thanks @ngimel .  I tried using the TORCH_LIBRARY macro and was able to get it working for a toy example.  However, the networks I'm testing have several custom kernels with different function signatures.   I'm getting build errors from the aten library for some of these kernels.  It looks like the compiler doesn't know how to interpret the function signature of the kernel I'm registering.

In general, I have to do performance testing on many networks that I didn't write and I'm seeing several examples of custom c++ kernels in these networks.  Registering each custom kernel is a lot of work and until all the kernels get registered the network graphs can't be generated.  The solution I proposed above makes it so the network sources would not need to be modified to get the desired outcome.  The proposed change just registers the output tensors produced by the custom kernels with the profiler record function.  This is very similar to the way the dispatcher registers the input and output tensors with the profiler.

**Comment by albanD** [comment]
There is already the proper tracer code included trace via `_trace_pre_record`/`_trace_post_record`.
What else would be missing for this?

**Comment by alexmsettle** [comment]
Interesting.  Do you have to have jit enabled to use the tracer?

The interface to the profiler I have been using is the RECORD_FUNCTION() macro here https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_function.cpp#L820

I added code to THPFunction_apply() that passes the outputs to the profiler after the outputs are processed.  The calling convention looks similar to _trace_pre_record / _trace_post_record.  In my case it is using [RecordFunction::setOutputs()  ](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/record_function.h#L339)

**Comment by albanD** [comment]
cc @robieta do you have more context on this?

**Comment by robieta** [comment]
I tend to agree with @ngimel. If you use the "run this opaque kernel" recipe it shouldn't be surprising that the framework just runs your code and that's it. If you want higher order features like profiling you have to use the proper full registration machinery. I actually think one of the coolest things about PyTorch is that all of our ops use the same machinery that we tell people to use for custom kernels, and thus it's possible for any custom kernel to have 100% feature parity with `aten` ops. So the idea of patching various bits of glue code to implement features that are already implemented elsewhere is not very appealing to us.

There are also benefits to living within the `TORCH_LIBRARY` path. That's the one we optimize, and what we look to when adding new features. (And there's a lot of semantic analysis capability that's being adding to profiler right now.)

**Comment by alexmsettle** [comment]
Thanks for the feedback @robieta @ngimel @albanD.  Sorry I didn't get back sooner, I have been working on other issues recently.

The topic of being able to use the profiler to capture an accurate network graph keeps coming up in my work.  Most recently it's showing up with NvFuser.

Let me try a simpler explanation for what I want to do.

The [RECORD_FUNCTION()](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/record_function.h#L564) macro is defined in the ATen lib in [record_function.h](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/record_function.h#L564).

Developers can use the macro to enable the logging of profiler events which will show up in the NVTX markers on GPU enabled systems.  However, this macro only lets you capture the name, input tensors and sequence_id of the function you're calling.  For the purposes of building a network graph it is also important to be able to capture the output tensors.

I would like to add a new macro that lets you record the outputs.  The profiler is already configured to manage output tensors if they are recorded.

`#define RECORD_OUTPUTS(outputs) \`
  `if (guard.needsOutputs()) {   \`
   ` guard.setOutputs(           \`
     `   std::vector<c10::IValue>(outputs.begin(), outputs.end()));\`
`  }`


This new macro would be used with RECORD_FUNCTION() like this.

`RECORD_FUNCTION("anna_fused_kernel",`
       ` std::vector<c10::IValue>  (inputs.begin(), inputs.end()), seq_id);`
`auto outputs = executor.runFusion(inputs, launch_params, input_id);`
`RECORD_OUTPUTS(outputs);`

This new macro uses existing profiler features and it allows users to visualize the dataflow across nodes in the network.  It can be used to solve my original question about custom kernels and for analyzing the different fusion engines in PyTorch.

**Comment by alexmsettle** [comment]
I just created a new issue for the RECORD_OUTPUTS() macro, I'll go ahead and close this issue.

https://github.com/pytorch/pytorch/issues/85575
